# Problem Statement
- This Dataset Contain the information about the bank customer that will take loan.
- Our goal is build a model that will detect weather the customer is eligible for loan or not.
- Data should be present in 2 different files.
- Both conatin same data about bank.
- Both files also contain `-99999` values.
- `-99999` are the missing values.

# Import Packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import warnings
warnings.filterwarnings('ignore')

# Read Data
1. Data is place another folder now find the path and read data.

In [2]:
file_path=[]
for i in os.listdir("Data"):
    file_path.append(f"{os.path.join(os.getcwd())}\Data\{i}")
    print(f"{os.path.join(os.getcwd())}\Data\{i}")

C:\Users\Laptop House\Github-Project\CreditRisk\Data\case_study1.xlsx
C:\Users\Laptop House\Github-Project\CreditRisk\Data\case_study2.xlsx


In [3]:
df1=pd.read_excel(file_path[0])
df2=pd.read_excel(file_path[1])

# Data Preprocessing
1. Top records
2. Shape of data
3. Check columns
4. Merge based on `common column`
5. Check Datatyes
6. Null values
7. Duplicates
8. Unique Values
9. Statistical Summary

# 1.Top Records

In [4]:
df1.sample(10)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
29420       29421         5              1              4   
4031         4032         1              1              0   
49299       49300         5              2              3   
5271         5272         7              1              6   
3798         3799         9              4              5   
50983       50984         2              1              1   
26379       26380         5              5              0   
1002         1003         8              3              5   
19019       19020         9              1              8   
45786       45787         1              0              1   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
29420                    0                  0            0.000   
4031                     0                  0            0.000   
49299                    0                  0            0.000   
5271                     0                  1            0.000   
3798                     4                  0            0.444   
50983                    0                  0            0.000   
26379                    0                  0            0.000   
1002                     1                  2            0.125   
19019                    6                  1            0.667   
45786                    0                  0            0.000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  CC_TL  \
29420              0.000          0.800          0.200  ...      0   
4031               0.000          0.000          1.000  ...      0   
49299              0.000          0.600          0.400  ...      1   
5271               0.143          0.857          0.143  ...      2   
3798               0.000          0.556          0.444  ...      0   
50983              0.000          0.500          0.500  ...      0   
26379              0.000          0.000          1.000  ...      0   
1002               0.250          0.625          0.375  ...      0   
19019              0.111          0.889          0.111  ...      1   
45786              0.000          1.000          0.000  ...      0   

       Consumer_TL  Gold_TL  Home_TL  PL_TL  Secured_TL  Unsecured_TL  \
29420            0        1        0      0           4             1   
4031             0        0        0      0           1             0   
49299            0        0        1      3           1             4   
5271             3        0        0      1           0             7   
3798             1        3        1      1           5             4   
50983            1        0        0      0           1             1   
26379            5        0        0      0           0             5   
1002             5        0        0      1           2             6   
19019            0        6        0      0           7             2   
45786            0        0        0      1           0             1   

       Other_TL  Age_Oldest_TL  Age_Newest_TL  
29420         1            117              9  
4031          1            272            272  
49299         0            164              7  
5271          1             17              6  
3798          2            106              2  
50983         0             48              6  
26379         0             65             24  
1002          0             41              5  
19019         1             23              2  
45786         0             13             13  

[10 rows x 26 columns]

In [5]:
df2.sample(10)

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
43028       43029                        117                           25   
6005         6006                       1937                           31   
39657       39658                         62                           35   
9861         9862                         42                       -99999   
40163       40164                        271                           35   
47226       47227                        335                           32   
37159       37160                        136                       -99999   
34265       34266                         73                       -99999   
46437       46438                        112                       -99999   
2047         2048                        237                       -99999   

       time_since_recent_deliquency  num_times_delinquent  \
43028                            22                     4   
6005                             19                    13   
39657                            29                     6   
9861                         -99999                     0   
40163                            15                     9   
47226                            31                     2   
37159                        -99999                     0   
34265                        -99999                     0   
46437                        -99999                     0   
2047                         -99999                     0   

       max_delinquency_level  max_recent_level_of_deliq  num_deliq_6mts  \
43028                     24                         23               0   
6005                     900                        900               0   
39657                    120                         60               0   
9861                  -99999                          0               0   
40163                    120                         60               0   
47226                     34                          9               0   
37159                 -99999                          0               0   
34265                 -99999                          0               0   
46437                 -99999                          0               0   
2047                  -99999                          0               0   

       num_deliq_12mts  num_deliq_6_12mts  ...  pct_CC_enq_L6m_of_L12m  \
43028                0                  0  ...                     0.0   
6005                 0                  0  ...                     0.0   
39657                0                  0  ...                     0.0   
9861                 0                  0  ...                     0.0   
40163                0                  0  ...                     0.0   
47226                0                  0  ...                     0.0   
37159                0                  0  ...                     1.0   
34265                0                  0  ...                     0.0   
46437                0                  0  ...                     0.0   
2047                 0                  0  ...                     0.0   

       pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  \
43028                     0.0                     0.0   
6005                      0.0                     0.0   
39657                     0.5                     0.0   
9861                      1.0                     0.0   
40163                     0.0                     0.0   
47226                     1.0                     0.0   
37159                     0.0                     1.0   
34265                     0.0                     0.0   
46437                     0.0                     0.0   
2047                      0.0                     0.0   

       max_unsec_exposure_inPct  HL_Flag  GL_Flag  last_prod_enq2  \
43028                     1.675        0        0              PL   
6005                 -99999.000        0        0          others   
39657      

# Conclussion
1. we can see that in `df2` most of the columns contain `-99999` values.
2. These `-99999` are basically `Missing value`.

# 2.shape

In [6]:
print("Data 1 Shape: ",df1.shape)
print("Data 2 Shape: ",df2.shape)

Data 1 Shape:  (51336, 26)
Data 2 Shape:  (51336, 62)


# Conclussion 
1. Data1 contain `51336` rows and `26` Column
2. Data2 contain `51336` rows and `62` Column

# 3.Columns

In [7]:
print(df1.columns)
print(df2.columns)

Index(['PROSPECTID', 'Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL',
       'Total_TL_opened_L6M', 'Tot_TL_closed_L6M', 'pct_tl_open_L6M',
       'pct_tl_closed_L6M', 'pct_active_tl', 'pct_closed_tl',
       'Total_TL_opened_L12M', 'Tot_TL_closed_L12M', 'pct_tl_open_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL',
       'Consumer_TL', 'Gold_TL', 'Home_TL', 'PL_TL', 'Secured_TL',
       'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL', 'Age_Newest_TL'],
      dtype='object')
Index(['PROSPECTID', 'time_since_recent_payment',
       'time_since_first_deliquency', 'time_since_recent_deliquency',
       'num_times_delinquent', 'max_delinquency_level',
       'max_recent_level_of_deliq', 'num_deliq_6mts', 'num_deliq_12mts',
       'num_deliq_6_12mts', 'max_deliq_6mts', 'max_deliq_12mts',
       'num_times_30p_dpd', 'num_times_60p_dpd', 'num_std', 'num_std_6mts',
       'num_std_12mts', 'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt',
       'num_dbt_6mts', 'num_dbt_

# Conclussion
1. See that both `dataframe` one common column `PROSPECTID`.
2. We can `merge` them first and then perform other data preprocessing.

# 4.Merge `df1` anf `df2`

In [8]:
df1.head(2)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
0           1         5              4              1                    0   
1           2         1              0              1                    0   

   Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
0                  0              0.0                0.0            0.2   
1                  0              0.0                0.0            1.0   

   pct_closed_tl  ...  CC_TL  Consumer_TL  Gold_TL  Home_TL  PL_TL  \
0            0.8  ...      0            0        1        0      4   
1            0.0  ...      0            1        0        0      0   

   Secured_TL  Unsecured_TL  Other_TL  Age_Oldest_TL  Age_Newest_TL  
0           1             4         0             72             18  
1           0             1         0              7              7  

[2 rows x 26 columns]

In [9]:
df2.head(2)

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
0           1                        549                           35   
1           2                         47                       -99999   

   time_since_recent_deliquency  num_times_delinquent  max_delinquency_level  \
0                            15                    11                     29   
1                        -99999                     0                 -99999   

   max_recent_level_of_deliq  num_deliq_6mts  num_deliq_12mts  \
0                         29               0                0   
1                          0               0                0   

   num_deliq_6_12mts  ...  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
0                  0  ...                     0.0                     0.0   
1                  0  ...                     0.0                     0.0   

   pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct  HL_Flag  GL_Flag  \
0                     0.0                    13.333        1        0   
1                     0.0                     0.860        0        0   

   last_prod_enq2  first_prod_enq2  Credit_Score  Approved_Flag  
0              PL               PL           696             P2  
1    ConsumerLoan     ConsumerLoan           685             P2  

[2 rows x 62 columns]

In [10]:
final_df=df1.merge(df2,how='inner',on='PROSPECTID')
final_df.head(2)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
0           1         5              4              1                    0   
1           2         1              0              1                    0   

   Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
0                  0              0.0                0.0            0.2   
1                  0              0.0                0.0            1.0   

   pct_closed_tl  ...  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
0            0.8  ...                     0.0                     0.0   
1            0.0  ...                     0.0                     0.0   

   pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct  HL_Flag  GL_Flag  \
0                     0.0                    13.333        1        0   
1                     0.0                     0.860        0        0   

   last_prod_enq2  first_prod_enq2  Credit_Score  Approved_Flag  
0              PL               PL           696             P2  
1    ConsumerLoan     ConsumerLoan           685             P2  

[2 rows x 87 columns]

# Conclussion
1. we have a `final df` now we can further work on this `final_df`
2. We should drop  `PROSPECTID` column b/c we don't need more

In [11]:
final_df.drop(columns=['PROSPECTID'],inplace=True)

# 5. Datatypes

In [12]:
data_types=final_df.dtypes.to_dict()

In [13]:
pd.reset_option('display.max_rows')
pd.reset_option('display.max_columns')
pd.DataFrame(list(data_types.items()),columns=['Columns','Datatypes']).head(87)

Columns Datatypes
0              Total_TL     int64
1         Tot_Closed_TL     int64
2         Tot_Active_TL     int64
3   Total_TL_opened_L6M     int64
4     Tot_TL_closed_L6M     int64
..                  ...       ...
81              GL_Flag     int64
82       last_prod_enq2    object
83      first_prod_enq2    object
84         Credit_Score     int64
85        Approved_Flag    object

[86 rows x 2 columns]

In [14]:
# object colums need =['material_status','Education','gender','last_prod_loan','first_prod_loan','Approved_flag']
# int columns need =[All except object col]

In [15]:
final_df.select_dtypes('object').columns

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2', 'Approved_Flag'],
      dtype='object')

In [16]:
final_df.select_dtypes('number').columns

Index(['Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL', 'Total_TL_opened_L6M',
       'Tot_TL_closed_L6M', 'pct_tl_open_L6M', 'pct_tl_closed_L6M',
       'pct_active_tl', 'pct_closed_tl', 'Total_TL_opened_L12M',
       'Tot_TL_closed_L12M', 'pct_tl_open_L12M', 'pct_tl_closed_L12M',
       'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL', 'Consumer_TL', 'Gold_TL',
       'Home_TL', 'PL_TL', 'Secured_TL', 'Unsecured_TL', 'Other_TL',
       'Age_Oldest_TL', 'Age_Newest_TL', 'time_since_recent_payment',
       'time_since_first_deliquency', 'time_since_recent_deliquency',
       'num_times_delinquent', 'max_delinquency_level',
       'max_recent_level_of_deliq', 'num_deliq_6mts', 'num_deliq_12mts',
       'num_deliq_6_12mts', 'max_deliq_6mts', 'max_deliq_12mts',
       'num_times_30p_dpd', 'num_times_60p_dpd', 'num_std', 'num_std_6mts',
       'num_std_12mts', 'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt',
       'num_dbt_6mts', 'num_dbt_12mts', 'num_lss', 'num_lss_6mts',
       'num_lss_12mts',

# Conclussion
1. `Datatypes` should be in correct format.

# 6. Null Values
1. First we can make the copy of dataframe

In [17]:
copy_df=final_df.copy()

In [18]:
final_df.isnull().sum()

Total_TL               0
Tot_Closed_TL          0
Tot_Active_TL          0
Total_TL_opened_L6M    0
Tot_TL_closed_L6M      0
                      ..
GL_Flag                0
last_prod_enq2         0
first_prod_enq2        0
Credit_Score           0
Approved_Flag          0
Length: 86, dtype: int64

1. **We can't see missing values but in above we say that `-99999` shows the missing values**
2. **Check if the ratio of `-99999` is greater in the specific column remove the columns**

# Handling Missing Values

In [19]:
final_df[final_df['time_since_recent_payment']==-99999].shape[0]

4291

In [20]:
final_df.shape

(51336, 86)

1. **We have `51336` records we can set the treshold `10000` if any columns contain `-99999` more then 10000 we can drop the columns.** 
2. **Also check the ratio of containg `-99999`.**

In [21]:
col_to_drop=[]
for i in final_df.columns:
    temp=final_df[final_df[i]==-99999].shape[0]
    ratio= round(temp/len(final_df),3)*100
    if(temp>10000):
        print(f"{i} -> -99999 value is -> {temp} and ratio is {ratio}%")
        col_to_drop.append(i)

time_since_first_deliquency -> -99999 value is -> 35949 and ratio is 70.0%
time_since_recent_deliquency -> -99999 value is -> 35949 and ratio is 70.0%
max_delinquency_level -> -99999 value is -> 35949 and ratio is 70.0%
max_deliq_6mts -> -99999 value is -> 12890 and ratio is 25.1%
max_deliq_12mts -> -99999 value is -> 10832 and ratio is 21.099999999999998%
CC_utilization -> -99999 value is -> 47636 and ratio is 92.80000000000001%
PL_utilization -> -99999 value is -> 44435 and ratio is 86.6%
max_unsec_exposure_inPct -> -99999 value is -> 23178 and ratio is 45.1%


# Conclussion
1. `time_since_first_deliquency` contain `70.0%` missing data.
2. `time_since_recent_deliquency` contain `70.0%` missing data.
3. `max_delinquency_level` contain `70.0%` missing data.
4. `max_deliq_6mts` contain `25.1%` missing data.
5. `max_deliq_12mts` contain  `21.09% missing data.
6. `CC_utilization` conrain `92.8%` missing data.
7. `PL_utilization` contain `86.6%` missing data.
8. `max_unsec_exposure_inPct` contain  `45.1%` missing data.
- **We will drop these columns**

### Drop Columns 

In [22]:
final_df.drop(columns=col_to_drop,inplace=True)

In [23]:
final_df.shape

(51336, 78)

In [24]:
final_df.sample(10)

Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
24247         1              1              0                    0   
7426         11              5              6                    0   
45377        12              1             11                    1   
49395         2              1              1                    0   
35219         1              0              1                    1   
48484         4              1              3                    0   
27530         5              3              2                    0   
41082        12              9              3                    1   
9433          5              0              5                    3   
48940         2              2              0                    0   

       Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
24247                  1            0.000                1.0          0.000   
7426                   0            0.000                0.0          0.545   
45377                  0            0.083                0.0          0.917   
49395                  0            0.000                0.0          0.500   
35219                  0            1.000                0.0          1.000   
48484                  0            0.000                0.0          0.750   
27530                  0            0.000                0.0          0.400   
41082                  0            0.083                0.0          0.250   
9433                   0            0.600                0.0          1.000   
48940                  0            0.000                0.0          0.000   

       pct_closed_tl  Total_TL_opened_L12M  ...  pct_PL_enq_L6m_of_L12m  \
24247          1.000                     1  ...                     0.0   
7426           0.455                     2  ...                     1.0   
45377          0.083                     3  ...                     0.0   
49395          0.500                     0  ...                     0.0   
35219          0.000                     1  ...                     0.0   
48484          0.250                     2  ...                     0.0   
27530          0.600                     0  ...                     0.4   
41082          0.750                     1  ...                     0.0   
9433           0.000                     4  ...                     1.0   
48940          1.000                     0  ...                     0.0   

       pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  \
24247                     0.0                     0.0                     0.0   
7426                      0.0                     0.5                     0.0   
45377                     0.0                     0.0                     0.0   
49395                     0.0                     0.0                     0.0   
35219                     0.0                     0.0                     0.0   
48484                     0.0                     0.0                     0.0   
27530                     0.0                     0.4                     0.0   
41082                     0.0                     0.0                     0.0   
9433                      0.0                     1.0                     0.0   
48940                     0.0                     0.0                     0.0   

       HL_Flag  GL_Flag  last_prod_enq2  first_prod_enq2  Credit_Score  \
24247        0        0    ConsumerLoan     ConsumerLoan           672   
7426         1        0    ConsumerLoan               PL           666   
45377        0        0              AL           others           699   
49395        0        1    ConsumerLoan           others           691   
35219        0        0    ConsumerLoan     ConsumerLoan           663   
48484        0        0          others           others           658   
27530        0        0              PL           others           684   
41082        1        0    ConsumerLoan               CC

# Conclussion
- After removing columns which contain `-99999` values more then `10000` shape of the dataframe is `51336` rows and `79` columns.
- we can remove only those columns which contain `-99999` values more then `10000` but remember that `-99999` are also present in those columns which contain  `-99999` values less then `10000`.
# Point to remember
- we have 2 option for `handling the remaing missing values`.
    ## Remove Missing Values
    - first we can remove those value.
        - we can remove the missing values if we maintain the balance of data we can remove them.
    ## Fill Missing Values
    - Fill these values
        - Filling the missing values is a cretical thing but if we can loss lot of data after removing the missing values then we pick the filling missing values.

- **Try to remove rows that contain `-99999` value**

In [25]:
for i in final_df.columns:
    print(final_df[~(final_df[i]==-99999)].shape[0])
    final_df=final_df[~(final_df[i]==-99999)]

51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51296
51296
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064


# Conclussion
- Check the shape after removing.
- Also check the ration before removing and after removing

In [26]:
print(f"Shape of the data after removing -9999 values is: {final_df.shape}")

Shape of the data after removing -9999 values is: (42064, 78)


In [27]:
len(final_df)/len(copy_df)*100

81.93860059217702

# Conclussion
- we see that after removing the rows that contain `-99999` values we can maintain `80%` data.
- Now we sholud pick remove missing value and we will remove them.

In [28]:
# final_df.to_csv('credit.csv',index=False)

# Duplicates

In [29]:
final_df.duplicated().sum()

0

# Conclussion
- There is no duplicate found in this dataset.

# Unique Values

In [30]:
for i in final_df.describe(include='object'):
    print("Unique valeu in: ",i)
    
    print(final_df[i].unique())
    
    print("Total Unique values: ",final_df[i].nunique())
    
    print("-"*50)
    print()

Unique valeu in:  MARITALSTATUS
['Married' 'Single']
Total Unique values:  2
--------------------------------------------------

Unique valeu in:  EDUCATION
['12TH' 'GRADUATE' 'SSC' 'POST-GRADUATE' 'UNDER GRADUATE' 'OTHERS'
 'PROFESSIONAL']
Total Unique values:  7
--------------------------------------------------

Unique valeu in:  GENDER
['M' 'F']
Total Unique values:  2
--------------------------------------------------

Unique valeu in:  last_prod_enq2
['PL' 'ConsumerLoan' 'AL' 'CC' 'others' 'HL']
Total Unique values:  6
--------------------------------------------------

Unique valeu in:  first_prod_enq2
['PL' 'ConsumerLoan' 'others' 'AL' 'HL' 'CC']
Total Unique values:  6
--------------------------------------------------

Unique valeu in:  Approved_Flag
['P2' 'P1' 'P3' 'P4']
Total Unique values:  4
--------------------------------------------------



# Conclussion
- There are `2` unique values in  `MARITALSTATUS` i-e  `['Married' 'Single']`
- There are `7` unique values in  `EDUCATION` i-e `['12TH' 'GRADUATE' 'SSC' 'POST-GRADUATE' 'UNDER GRADUATE' 'OTHERS'
 'PROFESSIONAL']`
- There are `2` unique values in  `GENDER` i-e `['M' 'F']`
- There are `6` unique values in  `last_prod_enq2` i-e `['PL' 'ConsumerLoan' 'AL' 'CC' 'others' 'HL']`
- There are `6` unique values in  `first_prod_enq2` i-e `['PL' 'ConsumerLoan' 'others' 'AL' 'HL' 'CC']`
- There are `4` unique values in  `Approved_Flag` i-e `['P2' 'P1' 'P3' 'P4']`

# Statistical Summary

In [31]:
final_df.describe()

Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
count  42064.000000   42064.000000   42064.000000         42064.000000   
mean       5.262980       2.967383       2.295597             0.812643   
std        7.463383       6.141098       2.404086             1.383559   
min        1.000000       0.000000       0.000000             0.000000   
25%        1.000000       0.000000       1.000000             0.000000   
50%        3.000000       1.000000       2.000000             0.000000   
75%        6.000000       3.000000       3.000000             1.000000   
max      235.000000     216.000000      47.000000            27.000000   

       Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
count        42064.00000     42064.000000       42064.000000   42064.000000   
mean             0.48992         0.179032           0.097783       0.577452   
std              1.05892         0.278043           0.210957       0.366110   
min              0.00000         0.000000           0.000000       0.000000   
25%              0.00000         0.000000           0.000000       0.286000   
50%              0.00000         0.000000           0.000000       0.545000   
75%              1.00000         0.333000           0.100000       1.000000   
max             19.00000         1.000000           1.000000       1.000000   

       pct_closed_tl  Total_TL_opened_L12M  ...  pct_currentBal_all_TL  \
count   42064.000000          42064.000000  ...           42064.000000   
mean        0.422548              1.672142  ...               0.883693   
std         0.366110              2.249543  ...              40.622275   
min         0.000000              0.000000  ...               0.000000   
25%         0.000000              0.000000  ...               0.152000   
50%         0.455000              1.000000  ...               0.600000   
75%         0.714000              2.000000  ...               0.860000   
max         1.000000             34.000000  ...            6327.500000   

            CC_Flag       PL_Flag  pct_PL_enq_L6m_of_L12m  \
count  42064.000000  42064.000000            42064.000000   
mean       0.102962      0.193063                0.219169   
std        0.303913      0.394707                0.395100   
min        0.000000      0.000000                0.000000   
25%        0.000000      0.000000                0.000000   
50%        0.000000      0.000000                0.000000   
75%        0.000000      0.000000                0.000000   
max        1.000000      1.000000                1.000000   

       pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  \
count            42064.000000            42064.000000            42064.000000   
mean                 0.074833                0.195497                0.064186   
std                  0.250658                0.367414                0.225989   
min                  0.000000                0.000000                0.000000   
25%                  0.000000                0.000000                0.000000   
50%                  0.000000                0.000000                0.000000   
75%                  0.000000                0.000000                0.000000   
max                  1.000000                1.000000                1.000000   

            HL_Flag       GL_Flag  Credit_Score  
count  42064.000000  42064.000000  42064.000000  
mean       0.252235      0.056580    679.326336  
std        0.434300      0.231042     21.133619  
min        0.000000      0.000000    469.000000  
25%        0.000000      0.000000    668.000000  
50%        0.000000      0.000000    679.000000  
75%        1.000000      0.000000    690.000000  
max        1.000000      1.000000    809.000000  

[8 rows x 72 columns]

# Conclussion
- **We can't see clear information first we can do `Feature Engnering` and then we can see clear information.**

# Feature Engineering

- In Feature Engineering, we can reduce the dimensionality of data by applying some statistical tests.

## Hyphothesis Testing
- Hyphothesis test are the statistical test used to draw conclussion about poopulation based on sample data.
- Some statistical test are that are use in this dataset.

### Variance Influence Threshold
- Suitable for numerical input and numerical output.

### Chi-Square Test
- Explain more about the Chi-Square test here.

### ANOVA
- These tests are applied on categorical versus numerical columns. Note that categorical variables should have more than three categories.


# Saperate numerical and categorical col

In [32]:
cat_col=final_df.select_dtypes('object').columns
num_col=final_df.select_dtypes('number').columns

In [33]:
cat_col

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2', 'Approved_Flag'],
      dtype='object')

In [34]:
num_col

Index(['Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL', 'Total_TL_opened_L6M',
       'Tot_TL_closed_L6M', 'pct_tl_open_L6M', 'pct_tl_closed_L6M',
       'pct_active_tl', 'pct_closed_tl', 'Total_TL_opened_L12M',
       'Tot_TL_closed_L12M', 'pct_tl_open_L12M', 'pct_tl_closed_L12M',
       'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL', 'Consumer_TL', 'Gold_TL',
       'Home_TL', 'PL_TL', 'Secured_TL', 'Unsecured_TL', 'Other_TL',
       'Age_Oldest_TL', 'Age_Newest_TL', 'time_since_recent_payment',
       'num_times_delinquent', 'max_recent_level_of_deliq', 'num_deliq_6mts',
       'num_deliq_12mts', 'num_deliq_6_12mts', 'num_times_30p_dpd',
       'num_times_60p_dpd', 'num_std', 'num_std_6mts', 'num_std_12mts',
       'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt', 'num_dbt_6mts',
       'num_dbt_12mts', 'num_lss', 'num_lss_6mts', 'num_lss_12mts',
       'recent_level_of_deliq', 'tot_enq', 'CC_enq', 'CC_enq_L6m',
       'CC_enq_L12m', 'PL_enq', 'PL_enq_L6m', 'PL_enq_L12m',
       'time_sinc

# Variance Inflence Factore (VIF)
- Now we can apply the `VIF` on numerical columns to check the relationship of each column to every other col.
- If the 2 or more columns are highly corelated so we can take only one column so that i will reduce the `multicolarnarty`.
- If we can reduce multicolarty the column will automaticlally reduce.

# Apply VIF
- we can set the threshold ratio if the variance is greater then treshold we can drop the columns.
- if the variance is less then treshold we can keep the columns.

In [35]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [36]:
col_to_kept=[]
vif_data=final_df[num_col]
column_index=0
for i in range(num_col.shape[0]):
    # variance_inflation_factor(dataframe,index)
    vif_value=variance_inflation_factor(vif_data,column_index)
    print(i,"------->",vif_value)
    
    if(vif_value<6):
        col_to_kept.append(num_col[i])
        column_index+=1
    else:
        vif_data=vif_data.drop(columns=num_col[i])

0 -------> inf
1 -------> inf
2 -------> 11.320180023967996
3 -------> 8.363698035000336
4 -------> 6.520647877790928
5 -------> 5.149501618212625
6 -------> 2.611111040579735
7 -------> inf
8 -------> 1788.7926256209232
9 -------> 8.601028256477228
10 -------> 3.8328007921530785
11 -------> 6.099653381646727
12 -------> 5.581352009642766
13 -------> 1.985584353098778
14 -------> inf
15 -------> 4.809538302819343
16 -------> 23.270628983464636
17 -------> 30.595522588100053
18 -------> 4.384346405965583
19 -------> 3.064658415523423
20 -------> 2.898639771299251
21 -------> 4.377876915347324
22 -------> 2.2078535836958433
23 -------> 4.916914200506864
24 -------> 5.214702030064725
25 -------> 3.3861625024231476
26 -------> 7.840583309478997
27 -------> 5.255034641721438
28 -------> inf
29 -------> 7.380634506427232
30 -------> 1.421005001517573
31 -------> 8.083255010190323
32 -------> 1.624122752404011
33 -------> 7.257811920140003
34 -------> 15.59624383268298
35 -------> 1.825857047

In [37]:
len(col_to_kept)

39

# Conclussion
- After applying `variance inflence factor` we can reduce some column now the remaing `numerical column is 40`
- we see that after applying `VIF` remaing numerical col is 40 now check the relationship of remaing numerical col to target column.

# ANOVA
- It is a hypothesis test in this test we can check the relationship b/w numerical col to target col.
- Target  col should be categorical and must contain 3 or greater then 3 categories.
- In our case our target col is `Approved_Flag` which contain `4 classes`.
- We can set the threshold `0.05` if the p_value is greater then `0.05` we can drop these columns.
- `0.05` threshold are generaly industry standard.

In [38]:
from scipy.stats import f_oneway

In [39]:
cat_col[-1]

'Approved_Flag'

In [40]:
col_to_remaing=[]
for i in col_to_kept:
    current_col=final_df[i]
    targte_col=final_df[cat_col[-1]]
    
    # Saperat each group for preforming ANOVA 
    group1=[value for value,group in zip(current_col,targte_col) if group=='P1']
    group2=[value for value,group in zip(current_col,targte_col) if group=='P2']
    group3=[value for value,group in zip(current_col,targte_col) if group=='P3']
    group4=[value for value,group in zip(current_col,targte_col) if group=='P4']
    
    # Perform ANOVA
    f_stats,p_value=f_oneway(group1,group2,group3,group4)
    print(i,"------>",p_value)
    
    if(p_value<0.05):
        col_to_remaing.append(i)

pct_tl_open_L6M ------> 0.0
pct_tl_closed_L6M ------> 2.676086811190998e-52
Tot_TL_closed_L12M ------> 1.762685578230421e-99
pct_tl_closed_L12M ------> 4.1384568281483056e-49
Tot_Missed_Pmnt ------> 9.572404615942641e-210
CC_TL ------> 5.710261952999025e-57
Home_TL ------> 0.0
PL_TL ------> 7.991892389241125e-235
Secured_TL ------> 0.0
Unsecured_TL ------> 2.102384431982427e-118
Other_TL ------> 0.0
Age_Oldest_TL ------> 0.0
Age_Newest_TL ------> 1.0465965445912546e-88
time_since_recent_payment ------> 8.943912807416738e-81
max_recent_level_of_deliq ------> 6.699932727692095e-81
num_deliq_6_12mts ------> 2.0772229883600566e-59
num_times_60p_dpd ------> 3.3833576401764997e-59
num_std_12mts ------> 0.0
num_sub ------> 1.6985545450345584e-22
num_sub_6mts ------> 0.016579119048249723
num_sub_12mts ------> 0.001299484021242129
num_dbt ------> 3.135918708089289e-11
num_dbt_12mts ------> 0.005446087209172642
num_lss ------> 0.009340555562870471
num_lss_12mts ------> 0.354899152686773
recent_l

In [41]:
len(col_to_remaing)

37

# Conclussion
- After `ANOVA` we can get only 38 columns.

# Chie-Square
- These test are generally perform on categorical column.
- Note that target column category should be `>=3`.
- Now using `chie-square` test we can check the relationship b/w target col.

In [42]:
from scipy.stats import chi2_contingency,chisquare

In [43]:
drop_col=[]
for i in cat_col[:-1]:
    # Perform chie-square
    chie,p_value,_,_=chi2_contingency(pd.crosstab(final_df[i],final_df[cat_col[-1]]))
    
    print(i,"----->",p_value)
    if(p_value>0.05):
        drop_col.append(i)

MARITALSTATUS -----> 3.578180861038862e-233
EDUCATION -----> 2.6942265249737532e-30
GENDER -----> 1.907936100186563e-05
last_prod_enq2 -----> 0.0
first_prod_enq2 -----> 7.84997610555419e-287


In [44]:
drop_col

[]

# Conclussion
- See that all the columns `p_value` is less the `0.05` so we can accept all the columns.

In [45]:
col_to_remaing

['pct_tl_open_L6M',
 'pct_tl_closed_L6M',
 'Tot_TL_closed_L12M',
 'pct_tl_closed_L12M',
 'Tot_Missed_Pmnt',
 'CC_TL',
 'Home_TL',
 'PL_TL',
 'Secured_TL',
 'Unsecured_TL',
 'Other_TL',
 'Age_Oldest_TL',
 'Age_Newest_TL',
 'time_since_recent_payment',
 'max_recent_level_of_deliq',
 'num_deliq_6_12mts',
 'num_times_60p_dpd',
 'num_std_12mts',
 'num_sub',
 'num_sub_6mts',
 'num_sub_12mts',
 'num_dbt',
 'num_dbt_12mts',
 'num_lss',
 'recent_level_of_deliq',
 'CC_enq_L12m',
 'PL_enq_L12m',
 'time_since_recent_enq',
 'enq_L3m',
 'NETMONTHLYINCOME',
 'Time_With_Curr_Empr',
 'CC_Flag',
 'PL_Flag',
 'pct_PL_enq_L6m_of_ever',
 'pct_CC_enq_L6m_of_ever',
 'HL_Flag',
 'GL_Flag']

In [46]:
cat_col

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2', 'Approved_Flag'],
      dtype='object')

# Conclussion
- In `col to kept remaing` contain only numerical col.
- `cat-col` conatin  only categorical col.
- Both are saperate now first combine them into a single list.

In [47]:
col_to_remaing.extend(cat_col)

In [48]:
len(col_to_remaing)

43

- **we see that out of `88` columns we can get `44` column that met the conditions.**

In [49]:
final_df=final_df[col_to_remaing]
final_df.head()

pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  pct_tl_closed_L12M  \
0            0.000                0.0                   0               0.000   
1            0.000                0.0                   0               0.000   
2            0.125                0.0                   0               0.000   
4            0.000                0.0                   0               0.000   
5            0.000                0.0                   1               0.167   

   Tot_Missed_Pmnt  CC_TL  Home_TL  PL_TL  Secured_TL  Unsecured_TL  ...  \
0                0      0        0      4           1             4  ...   
1                0      0        0      0           0             1  ...   
2                1      0        0      0           2             6  ...   
4                0      0        0      0           3             0  ...   
5                0      0        0      0           6             0  ...   

   pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  HL_Flag  GL_Flag  \
0                   0.000                     0.0        1        0   
1                   0.000                     0.0        0        0   
2                   0.000                     0.0        1        0   
4                   0.000                     0.0        0        0   
5                   0.429                     0.0        1        0   

   MARITALSTATUS      EDUCATION  GENDER  last_prod_enq2  first_prod_enq2  \
0        Married           12TH       M              PL               PL   
1         Single       GRADUATE       F    ConsumerLoan     ConsumerLoan   
2        Married            SSC       M    ConsumerLoan           others   
4        Married  POST-GRADUATE       M              AL               AL   
5        Married           12TH       M    ConsumerLoan               PL   

   Approved_Flag  
0             P2  
1             P2  
2             P2  
4             P1  
5             P3  

[5 rows x 43 columns]

In [50]:
final_df.shape

(42064, 43)

# Conclussion
- After Peprocessing we will get `42064` out of `51000` and `44` columns out of `88`

# Model Building
- First we can saperate the input and output columns.
- Second we can saperate numerical and categorical columns.
- Third we can build a pipeline for some transformation.

# Saperate Input and output col

In [51]:
feature=final_df.drop(columns=['Approved_Flag'])
label=final_df['Approved_Flag']

In [52]:
feature.head(2)

pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  pct_tl_closed_L12M  \
0              0.0                0.0                   0                 0.0   
1              0.0                0.0                   0                 0.0   

   Tot_Missed_Pmnt  CC_TL  Home_TL  PL_TL  Secured_TL  Unsecured_TL  ...  \
0                0      0        0      4           1             4  ...   
1                0      0        0      0           0             1  ...   

   PL_Flag  pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  HL_Flag  GL_Flag  \
0        1                     0.0                     0.0        1        0   
1        0                     0.0                     0.0        0        0   

   MARITALSTATUS  EDUCATION  GENDER  last_prod_enq2  first_prod_enq2  
0        Married       12TH       M              PL               PL  
1         Single   GRADUATE       F    ConsumerLoan     ConsumerLoan  

[2 rows x 42 columns]

In [53]:
label.head()

0    P2
1    P2
2    P2
4    P1
5    P3
Name: Approved_Flag, dtype: object

In [54]:
label.value_counts().sort_index()

Approved_Flag
P1     4908
P2    25452
P3     6440
P4     5264
Name: count, dtype: int64

# Encode target Colummn

In [55]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder,StandardScaler,OrdinalEncoder

In [56]:
lb=LabelEncoder()
label=lb.fit_transform(label)

In [57]:
label

array([1, 1, 1, ..., 2, 1, 1])

# Saperate `Numerical` and `Categorical` column

In [58]:
num_col=feature.select_dtypes('number').columns
cat_col=feature.select_dtypes('object').columns

In [59]:
num_col

Index(['pct_tl_open_L6M', 'pct_tl_closed_L6M', 'Tot_TL_closed_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'CC_TL', 'Home_TL', 'PL_TL',
       'Secured_TL', 'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL',
       'Age_Newest_TL', 'time_since_recent_payment',
       'max_recent_level_of_deliq', 'num_deliq_6_12mts', 'num_times_60p_dpd',
       'num_std_12mts', 'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt',
       'num_dbt_12mts', 'num_lss', 'recent_level_of_deliq', 'CC_enq_L12m',
       'PL_enq_L12m', 'time_since_recent_enq', 'enq_L3m', 'NETMONTHLYINCOME',
       'Time_With_Curr_Empr', 'CC_Flag', 'PL_Flag', 'pct_PL_enq_L6m_of_ever',
       'pct_CC_enq_L6m_of_ever', 'HL_Flag', 'GL_Flag'],
      dtype='object')

In [60]:
cat_col

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2'],
      dtype='object')

# Train Test Split

In [61]:
from sklearn.model_selection import train_test_split,cross_val_score

In [62]:
x_train,x_test,y_train,y_test=train_test_split(feature,label,test_size=0.2,random_state=43)

In [63]:
x_train.shape

(33651, 42)

In [64]:
x_test.shape

(8413, 42)

# Building Pipeline
- we can build a pipeline for numercial column.
    - In numerical pipeline we can scale the feature.
- we can build a pipeline for categorical column.
    - In categorical pipeline we can encode the categorical column.

In [65]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

# Numerical Pipeline

In [66]:
num_pipe=Pipeline(steps=[
    ("impute",SimpleImputer(strategy='median')),
    ('scale',StandardScaler())
])
num_pipe

Pipeline(steps=[('impute', SimpleImputer(strategy='median')),
                ('scale', StandardScaler())])

# Categorical Pipeline

In [67]:
cat_pipe=Pipeline(steps=[
    ("impute",SimpleImputer(strategy='most_frequent')),
    ('Encode',OneHotEncoder(drop='first',sparse=False,handle_unknown='ignore'))
])
cat_pipe

Pipeline(steps=[('impute', SimpleImputer(strategy='most_frequent')),
                ('Encode',
                 OneHotEncoder(drop='first', handle_unknown='ignore',
                               sparse=False))])

# Colum Transformer
- In column transformer we can combine al the steps.

In [68]:
from sklearn.compose import ColumnTransformer

In [69]:
transformer=ColumnTransformer(transformers=[
    ('num_transformer',num_pipe,num_col),
    ("Encode",cat_pipe,cat_col)
],remainder='passthrough')
transformer

ColumnTransformer(remainder='passthrough',
                  transformers=[('num_transformer',
                                 Pipeline(steps=[('impute',
                                                  SimpleImputer(strategy='median')),
                                                 ('scale', StandardScaler())]),
                                 Index(['pct_tl_open_L6M', 'pct_tl_closed_L6M', 'Tot_TL_closed_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'CC_TL', 'Home_TL', 'PL_TL',
       'Secured_TL', 'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL',...
       'Time_With_Curr_Empr', 'CC_Flag', 'PL_Flag', 'pct_PL_enq_L6m_of_ever',
       'pct_CC_enq_L6m_of_ever', 'HL_Flag', 'GL_Flag'],
      dtype='object')),
                                ('Encode',
                                 Pipeline(steps=[('impute',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('Encode',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore',
                                                                sparse=False))]),
                                 Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2'],
      dtype='object'))])

# Build Final Pipeline for model

# Fit the model
- In this we can train different models.
- Those model will give better result we can tune them.

In [70]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
import xgboost as xgb
from xgboost import XGBClassifier

# Make the models dict

In [71]:
model_dic={
    "LogisticRegression":LogisticRegression(),
    "RandomForest":RandomForestClassifier(),
    "DecessionTree":DecisionTreeClassifier(),
    "xgboost":XGBClassifier()
}

# Fit models
- Fit models and also calculate the score

In [72]:
from sklearn.metrics import accuracy_score,confusion_matrix

In [73]:
results={
    "model_name":[],
    "score":[],
    'train score':[],
    "test score":[],
    'matrix':[]
}
for model_name,model in model_dic.items():
    final=Pipeline(steps=[
    ("process",transformer),
    ("model",model)
    ])
    
    # Fit the model pipe
    final.fit(x_train,y_train)
    
    # Prediction
    pre=final.predict(x_test)
    
    # Calculate score
    score=accuracy_score(y_test,pre)
    matrix=confusion_matrix(y_test,pre)
    
    #Cross val score
    train_score=cross_val_score(final,x_train,y_train,cv=5,scoring='accuracy')
    test_score=cross_val_score(final,x_test,y_test,cv=5,scoring='accuracy')
    
    
    results['model_name'].append(model_name)
    results['score'].append(score)
    results['train score'].append(train_score)
    results['test score'].append(test_score)
    
    results['matrix'].append(matrix)

# See the results

In [74]:
pd.DataFrame(results)

model_name     score  \
0  LogisticRegression  0.754903   
1        RandomForest  0.772257   
2       DecessionTree  0.707001   
3             xgboost  0.783193   

                                         train score  \
0  [0.7517456544347051, 0.7526002971768202, 0.747...   
1  [0.7704650126281385, 0.7655274888558692, 0.765...   
2  [0.6969246768682217, 0.6925705794947994, 0.695...   
3  [0.7715049769722181, 0.7717682020802378, 0.773...   

                                          test score  \
0  [0.7498514557338086, 0.7486631016042781, 0.742...   
1  [0.7658942364824718, 0.7712418300653595, 0.767...   
2  [0.6892453951277481, 0.6939988116458705, 0.691...   
3  [0.7748068924539513, 0.7712418300653595, 0.775...   

                                              matrix  
0  [[679, 286, 2, 3], [145, 4847, 86, 74], [28, 8...  
1  [[702, 268, 0, 0], [130, 4791, 182, 49], [30, ...  
2  [[670, 253, 44, 3], [244, 4172, 611, 125], [31...  
3  [[759, 210, 1, 0], [145, 4714, 248, 45], [34, ...

# Confussion Matrix

## Logistic Regression 

In [75]:
results['matrix'][0]

array([[ 679,  286,    2,    3],
       [ 145, 4847,   86,   74],
       [  28,  864,  158,  217],
       [   0,  262,   95,  667]], dtype=int64)

# Random Forest

In [76]:
results['matrix'][1]

array([[ 702,  268,    0,    0],
       [ 130, 4791,  182,   49],
       [  30,  756,  257,  224],
       [   2,  142,  133,  747]], dtype=int64)

# Desession Tree

In [77]:
results['matrix'][2]

array([[ 670,  253,   44,    3],
       [ 244, 4172,  611,  125],
       [  31,  538,  439,  259],
       [   2,  133,  222,  667]], dtype=int64)

# XgBoost

In [78]:
results['matrix'][3]

array([[ 759,  210,    1,    0],
       [ 145, 4714,  248,   45],
       [  34,  644,  381,  208],
       [   0,   94,  195,  735]], dtype=int64)

# Conclussion
- We see that `Xgboost` give better result with `accuracy score` of `78` now we can fine tune them

In [79]:
pd.Series(y_test).value_counts().sort_index()

0     970
1    5152
2    1267
3    1024
Name: count, dtype: int64

# Hyper Parameter Tunning

In [80]:
from sklearn.model_selection import GridSearchCV

In [81]:
param_grid = {
    'model__model__colsample_bytree': [0.1, 0.3, 0.5, 0.7, 0.9],
    'model__model__learning_rate': [0.001, 0.01, 0.1, 1],
    'model__model__max_depth': [3, 5, 8, 10],
    'model__model__n_estimators': [10, 50, 100]
}

In [82]:
xgb_boost=XGBClassifier()
pipeline=Pipeline(steps=[
    ("process",transformer),
    ('model',xgb_boost)
])

In [83]:
grid_search=GridSearchCV(pipeline,param_grid=param_grid,scoring='accuracy',n_jobs=-1,cv=5)
grid_search.fit(x_train,y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('process',
                                        ColumnTransformer(remainder='passthrough',
                                                          transformers=[('num_transformer',
                                                                         Pipeline(steps=[('impute',
                                                                                          SimpleImputer(strategy='median')),
                                                                                         ('scale',
                                                                                          StandardScaler())]),
                                                                         Index(['pct_tl_open_L6M', 'pct_tl_closed_L6M', 'Tot_TL_closed_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'CC_TL', 'Home_TL', 'PL_TL',...
                                                      missing=nan,
                                                      monotone_constraints=None,
                                                      multi_strategy=None,
                                                      n_estimators=None,
                                                      n_jobs=None,
                                                      num_parallel_tree=None,
                                                      random_state=None, ...))]),
             n_jobs=-1,
             param_grid={'model__model__colsample_bytree': [0.1, 0.3, 0.5, 0.7,
                                                            0.9],
                         'model__model__learning_rate': [0.001, 0.01, 0.1, 1],
                         'model__model__max_depth': [3, 5, 8, 10],
                         'model__model__n_estimators': [10, 50, 100]},
             scoring='accuracy')

In [84]:
grid_search.best_params_

{'model__model__colsample_bytree': 0.1,
 'model__model__learning_rate': 0.001,
 'model__model__max_depth': 3,
 'model__model__n_estimators': 10}

In [85]:
pre=grid_search.predict(x_test)
pre

array([1, 2, 1, ..., 3, 1, 1], dtype=int64)

# Check accuracy again

In [86]:
accuracy_score(y_test,pre)

0.7831926779983359

# confussion Matrix

In [87]:
confusion_matrix(y_test,pre)

array([[ 759,  210,    1,    0],
       [ 145, 4714,  248,   45],
       [  34,  644,  381,  208],
       [   0,   94,  195,  735]], dtype=int64)

# Corss Validation

In [88]:
train_score=cross_val_score(pipeline,x_train,y_train,cv=5,scoring='accuracy').mean()
test_score=cross_val_score(pipeline,x_test,y_test,cv=5,scoring='accuracy').mean()

In [89]:
print("Train score is: ",train_score)
print("Test Score is: ",test_score)

Train score is:  0.77379579480009
Test Score is:  0.772732288966464
